main.go

```diff
@@ -21,97 +22,62 @@ func main() {
+	spanMinLimit := flag.Int("sm", 1, "min spans per transaction")
+
 	flag.Parse()
 
 	if *spanMaxLimit < *spanMinLimit {
 		spanMaxLimit = spanMinLimit
 	}
-
-	// configure tracer
-	logger := newApmLogger(log.New(os.Stderr, "", log.Ldate|log.Ltime|log.Lshortfile))
-	rand.Seed(*seed)
-	logger.Debugf("random seed: %d", *seed)
-
-	tracer := apm.DefaultTracer
-	flush := func() {
-		flushed := make(chan struct{})
-		go func() {
-			tracer.Flush(nil)
-			close(flushed)
-		}()
-
-		flushWait := time.After(*flushTimeout)
-		if *flushTimeout == 0 {
-			flushWait = make(<-chan time.Time)
-		}
-		select {
-		case <-flushed:
-		case <-flushWait:
-			// give up waiting for flush
-			logger.Errorf("timed out waiting for flush to complete")
-		}
-		tracer.Close()
-	}
-	tracer.SetLogger(logger)
-	tracer.SetMetricsInterval(0) // disable metrics
-	transport := tracer.Transport.(*apmtransport.HTTPTransport)
-	transport.SetUserAgent("hey-apm")
-	if *apmServerSecret != "" {
-		transport.SetSecretToken(*apmServerSecret)
-	}
-	if *apmServerUrl != "" {
-		u, err := url.Parse(*apmServerUrl)
-		if err != nil {
-			logger.Fatal("invalid apm server url:", err)
-		}
-		transport.SetServerURL(u)
+	if *errorFrameMaxLimit < *errorFrameMinLimit {
+		errorFrameMaxLimit = errorFrameMinLimit
 	}
 
-	tracer.SetSpanFramesMinDuration(1 * time.Nanosecond)
-	tracer.SetMaxSpans(*spanMaxLimit)
+	logger := out.NewApmLogger(log.New(os.Stderr, "", log.Ldate|log.Ltime|log.Lshortfile))
+	rand.Seed(*seed)
 
-	w := worker{
-		apmLogger:  logger,
-		Tracer:     tracer,
-		runTimeout: *runTimeout,
+	w := work.Worker{
+		RunTimeout:             *runTimeout,
+		TransactionLimit:       *transactionLimit,
+		TransactionFrequency:   *transactionFrequency,
+		MaxSpansPerTransaction: *spanMaxLimit,
+		MinSpansPerTransaction: *spanMinLimit,
+		ErrorLimit:             *errorLimit,
+		ErrorFrequency:         *errorFrequency,
+		MaxFramesPerError:      *errorFrameMaxLimit,
+		MinFramesPerError:      *errorFrameMinLimit,
 	}
-	w.addErrors(throttle(time.NewTicker(*errorFrequency).C), *errorLimit, *errorFrameMinLimit, *errorFrameMaxLimit)
-	w.addTransactions(throttle(time.NewTicker(*transactionFrequency).C), *transactionLimit, *spanMinLimit, *spanMaxLimit)
+	w.Add(work.HandleSignals())
 
 	logger.Debugf("start")
 	defer logger.Debugf("finish")
-	report, _ := w.Work()
-	defer func() { logger.Debugf("%s elapsed since event generation completed", time.Now().Sub(report.End)) }()
-	defer flush()
-	e, t, s := w.Counts()
-	logger.Printf("generated %d events (errors: %d, transctions: %d, spans: %d) in %s",
-		e+t+s, e, t, s, report.End.Sub(report.Start))
-	if report.Count != t {
-		logger.Errorf("unexpected sampling decision count, expected: %d got: %d", t, report.Count)
-	}
+	tracer := tracer.NewTracer(logger, *flushTimeout, *apmServerSecret, *apmServerUrl)
+
+	report, _ := w.Work(tracer)
+	logger.Debugf("%s elapsed since event generation completed", time.Now().Sub(report.End))
```

Comment:
this isn't accurate because `Report.End` is [recorded](https://github.com/elastic/hey-apm/pull/100/files#diff-6ddb7b29d9aadc915e46a2996ebe1b15R63) after flush.  How about adding a `report.Stop` and using that for this calculation instead?  I think it's important as `-run 5s -flush 0` may take much longer than 5s and this explains why

Reply:
good point